File: scanlibrary/src/main/java/com/scanlibrary/ScanActivity.java
package com.scanlibrary;

import android.content.ComponentCallbacks2;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.transition.Fade;
import android.view.Window;

import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;


/**
 * Created by jhansi on 28/03/15.
 */
public class ScanActivity extends AppCompatActivity implements IScanner, ComponentCallbacks2 {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // inside your activity (if you did not enable transitions in your theme)
        getWindow().requestFeature(Window.FEATURE_CONTENT_TRANSITIONS);

        // set an exit transition
        Fade fade = new Fade();
        fade.setDuration(1000);

        getWindow().setEnterTransition(fade);
        getWindow().setExitTransition(fade);

        setContentView(R.layout.scan_layout);
        init();
        Utils.setLightNavigationBar(findViewById(android.R.id.content), this);
    }

    private void init() {
        PickImageFragment fragment = new PickImageFragment();
        Bundle bundle = new Bundle();
        bundle.putInt(ScanConstants.OPEN_INTENT_PREFERENCE, getPreferenceContent());
        fragment.setArguments(bundle);
        getSupportFragmentManager()
                .beginTransaction()
                .add(R.id.content, fragment)
                .commit();
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();

        final android.app.FragmentManager fragmentManager = getFragmentManager();
        fragmentManager.addOnBackStackChangedListener(() -> {
            if (fragmentManager.getBackStackEntryCount() <= 1) {
                finish();
            }
        });

    }

    protected int getPreferenceContent() {
        return getIntent().getIntExtra(ScanConstants.OPEN_INTENT_PREFERENCE, 0);
    }

    @Override
    public void onBitmapSelect(Uri uri) {
        Fragment fragment = new ScanFragment();
        Bundle bundle = new Bundle();
        bundle.putParcelable(ScanConstants.SELECTED_BITMAP, uri);
        fragment.setArguments(bundle);
        getSupportFragmentManager()
                .beginTransaction()
                .add(R.id.content, fragment)
                .commit();
    }

    @Override
    public void onScanFinish(Uri uri) {
        ImageFiltersFragment fragment = new ImageFiltersFragment();
        Bundle bundle = new Bundle();
        bundle.putParcelable(ScanConstants.SCANNED_RESULT, uri);
        fragment.setArguments(bundle);

        getSupportFragmentManager()
                .beginTransaction()
                .add(R.id.content, fragment)
                .addToBackStack(ImageFiltersFragment.class.toString())
                .commit();
    }

    public native Bitmap getScannedBitmap(Bitmap bitmap, float x1, float y1, float x2, float y2, float x3, float y3, float x4, float y4);

    public native Bitmap getGrayBitmap(Bitmap bitmap);

    public native Bitmap getMagicColorBitmap(Bitmap bitmap);

    public native Bitmap getBWBitmap(Bitmap bitmap);

//    public native float[] getPoints(Bitmap bitmap);

    static {
        System.loadLibrary("opencv_java3");
        System.loadLibrary("Scanner");
    }
}

File: scanlibrary/src/main/java/com/scanlibrary/ScanActivity.kt
package com.scanlibrary

import android.content.ComponentCallbacks2
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.net.Uri
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment

/**
 * Created by jhansi on 28/03/15.
 */
class ScanActivity : AppCompatActivity(), IScanner, ComponentCallbacks2 {

    companion object {
        init {
            System.loadLibrary("opencv_java3")
            System.loadLibrary("Scanner")
        }

        private const val EXTRA_IMAGE_URI = "extra_image_uri"
        const val EXTRA_RESUL_IMAGE_NAME = "extra_resul_image_name"

        fun makeIntent(context: Context, uri: Uri) =
            Intent(context, ScanActivity::class.java).apply {
                putExtra(EXTRA_IMAGE_URI, uri)
            }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.scan_layout)
        init(intent.extras?.getParcelable(EXTRA_IMAGE_URI)!!)
    }

    private fun init(uri: Uri) {
        val fragment: Fragment = ScanFragment()
        val bundle = Bundle()
        bundle.putParcelable(ScanConstants.SELECTED_BITMAP, uri)
        fragment.arguments = bundle
        supportFragmentManager
            .beginTransaction()
            .add(R.id.content, fragment)
            .commit()
    }

    override fun onScanFinish(uri: Uri) {
        val fragment = ImageFiltersFragment()
        val bundle = Bundle()
        bundle.putParcelable(ScanConstants.SCANNED_RESULT, uri)
        fragment.arguments = bundle
        supportFragmentManager
            .beginTransaction()
            .add(R.id.content, fragment)
            .addToBackStack(ImageFiltersFragment::class.java.toString())
            .commit()
    }

    external override fun getScannedBitmap(
        bitmap: Bitmap,
        x1: Float,
        y1: Float,
        x2: Float,
        y2: Float,
        x3: Float,
        y3: Float,
        x4: Float,
        y4: Float
    ): Bitmap

    external override fun getGrayBitmap(bitmap: Bitmap): Bitmap
    external override fun getMagicColorBitmap(bitmap: Bitmap): Bitmap
    external override fun getBWBitmap(bitmap: Bitmap): Bitmap


}
